iphone/cli/lib/process-js-task.js

```diff
@@ -0,0 +1,330 @@
+
+		// @fixme: can be removed in 9.0 to even further decrease build times on incremental builds
+		const unchangedFiles = Array.from(this.inputFiles).filter(filePath => !changedFiles.has(filePath));
+		const unchangedPromise = Promise.all(unchangedFiles.map(filePath => this.processJsFile(filePath)));
+
+		return Promise.all([ deletedPromise, updatedPromise, unchangedPromise ]);
+	}
+
+	/**
+	 * Loads last run's task data if this task will be skipped entirely
+	 *
+	 * @return {Promise}
+	 */
+	loadResultAndSkip() {
+		this.loadTaskData();
+
+		return Promise.resolve();
+	}
+
+	/**
+	 * Funktion that will be called after the task run finished.
+	 *
+	 * Used to populate values back into the builder and update the task data
+	 * before it gets written to disk.
+	 */
+	afterTaskAction() {
+		this.builder.tiSymbols = this.data.tiSymbols;
+
+		this.data.jsBootstrapFiles = this.jsBootstrapFiles;
+		this.data.jsFiles = this.jsFiles;
+
+		this.saveTaskData();
+
+		this.builder.unmarkBuildDirFiles(this.incrementalDirectory);
+	}
+
+	/**
+	 * Processes a single JavaScript file by applying several transforms and then copying it to
+	 * its destination path.
+	 *
+	 * If the content of the unprocessed JavaScript file did not change compared to the last run,
+	 * any further processing will be skipped.
+	 *
+	 * @param {String} filePathAndName Full path to the JavaScript file
+	 * @return {Promise}
+	 */
+	processJsFile(filePathAndName) {
+		let file = this.resolveRelativePath(filePathAndName);
+		if (!file) {
+			throw new Error(`Unable to resolve relative path for ${filePathAndName}.`);
+		}
+
+		const bootstrapPath = file.substr(0, file.length - 3);  // Remove the ".js" extension.
+		if (bootstrapPath.endsWith('.bootstrap') && !this.jsBootstrapFiles.includes(bootstrapPath)) {
+			this.jsBootstrapFiles.push(bootstrapPath);
+		}
+
+		const info = this.jsFiles[file];
+		if (this.builder.encryptJS) {
+			if (file.indexOf('/') === 0) {
+				file = path.basename(file);
+			}
+			this.builder.jsFilesEncrypted.push(file); // original name
+			file = file.replace(/\./g, '_');
+			info.dest = path.join(this.buildAssetsDir, file);
+			this.builder.jsFilesToEncrypt.push(file); // encrypted name
+		}
+
+		return new Promise((resolve) => {
+			try {
+				this.builder.cli.createHook('build.ios.copyResource', this.builder, (from, to, cb) => {
+					const originalContents = fs.readFileSync(from).toString();
+
+					const r = {
+						original: originalContents,
+						contents: originalContents,
+						symbols: []
+					};
+
+					this.builder.cli.createHook('build.ios.compileJsFile', this.builder, (r, from, to, cb2) => {
+						// Read the possibly modified file contents
+						const source = r.contents;
+
+						// If the file didn't change from previous run, return early
+						const currentHash = this.generateHash(source);
+						const previousHash = this.data.contentHashes[from];
+						if (previousHash && previousHash === currentHash) {
```

Comment:
This is basically where all the magic happens. If the source file content did not change BEFORE we run `jsanalyze` just skip the whole thing, re-use the existing file and load used Ti symbols from previous run.

Reply:
We cannot skip `jsanalyze`. We need to AST parse everything so we can detect all `Ti.*` and `Titanium.*` symbols and enable the various APIs such as Ti.Media. Same goes for Android (and maybe Windows?).

I highly recommend testing using iOS device builds of KitchenSink. There's a lot of sensitive code here.

Reply:
I store the symbols at the end so they can be re-used on the next run. If nothing in the source file changed i don't see why we would need to run the `jsanalyze` again. We can simply [load the symbols](https://github.com/appcelerator/titanium_mobile/pull/10805/files#diff-597ce7d1e6d7ef73f52de3af52512094R309) from the last run and inject them back [into the builder](https://github.com/appcelerator/titanium_mobile/pull/10805/files#diff-597ce7d1e6d7ef73f52de3af52512094R98).

I tested this with our KitchenSink app using device builds and this works just fine. The symbols for unchanged files will be re-used from the last build and changed files will run `jsanalyze` again and have their Ti symbols updated. If a file gets removed then it's stored Ti symbols also get cleared.

Reply:
So I think the concern here is that we need to take into account more than just the original source file changing. Did any options baked into jsanalyze change? minify, transpile, etc. Basically anything that goes into `analyzeOptions`. Additionally, and this is implicit and not properly handled as-is, but the user could have a .babelrc/config that will get loaded any modify babel usage. I think @brentonhouse is doing this. We haven't officially supported/fixed handling that yet, but it's an expected use case. (right now I think for example we bomb out if a babel config ignores a file because transform returns null and we expect an object with a `code` property).

Second, is obviously we are generating both a transformed file and gathering Ti API usage. You need to handle "storing" both of those and handling merging the Ti API usage symbols stuff back into the build properly. I don't think that's happening correctly here.

Reply:
The `jsanalyze` options are a good point. I can check those before the task starts and fallback to a full run if anything changed. I can also feed in possible .babelrc/config paths and force a full rebuild if any of those changed. So that's also not a problem once that is officially supported.

Regarding your second point, I'm doing exactly that. The TI API usage symbols are stored per file in the task data that get's written to disk and will be loaded and merged back into the builder on subsequent runs. The transformed file is preserved between builds anyway so it's also kind of "stored".